### PR TITLE
refactor: major code refactor

### PR DESCRIPTION
**Pull request by fisker**
Refactor sortSubKey logic, prepare for #83 to ignore some field.

I'm planning to move all this to sortOrder, only problem is `compareScriptKeys` is in the inner scope, will try something else later

### CODE REVIEW

**Comment by keithamus** [comment]
To fix #83 we can just avoid sorting `keywords`. Is there another justification for this refactor?

**Comment by fisker** [comment]
I'm trying to make it easier to add new field,

we don't need add `mocha` to `sortOrder`, then add a new line `sortSubKey('mocha')`,  we can add `{field: 'mocha', sort: true}` something like this

**Comment by fisker** [comment]
double checked, I didn't miss any field

**Comment by keithamus** [comment]
I think it might be useful to minimise. Let's see a complete implementation then, that combines `sortOrder` with these subkeys.

**Comment by fisker** [comment]
I look at #83 again,  only ignore `keywords` from sorting, might don't need this.
I thought it ask for ignore some fields. 

But I think still useful to put all this together.

One question, should we support shortcut `"key"` for `{key, sort: true}`(28 of them) or `{key, sort: false}`(31 of them)

File: index.js [hunk 1]
```diff
@@ -119,15 +130,31 @@ function sortPackageJson(packageJson, options = {}) {
...
       return;
     }
+
     if (typeof packageJson[key] === 'object') {
+      let sortList
+      if (sortScripts) {
+        sortList = compareScriptKeys
+      } else if (Array.isArray(sort)) {
+        sortList = sort
+      }
```
Comment (keithamus) [inline index.js]: This logic seems a little complicated. Can we just provide `sortList` as an option to the function?

File: index.js [hunk 2]
```diff
@@ -3,76 +3,84 @@ const sortObjectKeys = require('sort-object-keys');
...
+  {key: 'version', sortSubKey: false},
+  {key: 'private', sortSubKey: true},
+  {key: 'description', sortSubKey: false},
+  {key: 'keywords', sortSubKey: true, unique: true},
+  {key: 'homepage', sortSubKey: true},
+  {key: 'bugs', sortSubKey: true, sortList: ['url', 'email']},
+  {key: 'repository', sortSubKey: true, sortList: ['type', 'url']},
+  {key: 'funding', sortSubKey: true, sortList: ['type', 'url']},
+  {key: 'license', sortSubKey: true, sortList: ['type', 'url']},
+  {key: 'author', sortSubKey: true, sortList: ['name', 'email', 'url']},
```
Comment (fisker) [inline index.js]: Do you think this is a little complicated?

I was trying to make this simplier like this


```js
{
  'name',
  'version',
  {key: 'homepage'},
  {key: 'bugs', sortList: ['url', 'email']}
}
```

But, seems not very clear that `homepage` and `bugs` will sort sub key, any suggestion?

**Comment by keithamus** [comment]
@fisker I've done a bit more refactoring, cleaned up a lot of stuff. Let me know if you think this is good?

File: index.js [hunk 3]
```diff
@@ -85,183 +24,245 @@ const defaultNpmScripts = [
...
+  { key: 'private' },
+  { key: 'description' },
+  { key: 'keywords', over: uniq },
+  { key: 'homepage' },
+  { key: 'bugs', over: sortObjectBy(['url', 'email']) },
+  { key: 'repository', over: sortURLObject },
+  { key: 'funding', over: sortURLObject },
+  { key: 'license', over: sortURLObject },
+  { key: 'author', over: sortAuthorObject },
+  { key: 'files' },
```
Comment (fisker) [inline index.js]: I think this was sorted before, as array, maybe we can also unique it

**Comment by fisker** [comment]
Looks great,  I'll check again when I get back to my laptop

File: package.json [hunk 4]
```diff
@@ -35,14 +36,35 @@
       "commit-msg": "validate-commit-msg"
     }
   },
+  "prettier": {
+    "semi": false,
+    "singleQuote": true,
+    "trailingComma": "all"
```
Comment (fisker) [inline package.json]:
```suggestion
    "trailingComma": "es5"
```

File: index.js (hunk 3 above)
Comment (fisker) [reply index.js]: sorry, it's not sorted before,  in this case

```js
  "files": [
    "index.js",
    "index.d.ts"
  ],
```

I guess , it doesn't make sense to put `index.d.ts` above `index.js`, wo we should keep the order as it is.

add uniq in https://github.com/keithamus/sort-package-json/pull/90/commits/e10cf9ad3389d13e27a0e617419718db6a7ae78a

**Comment by fisker** [comment]
@keithamus I did a little change, let me know if there is any problem

**Comment by fisker** [comment]
You did great  job on this refactoring, like to see it get merged.

File: index.js [hunk 5]
```diff
@@ -85,183 +27,209 @@ const defaultNpmScripts = [
...
-  let hasWindowsNewlines = false;
-  let endCharacters = '';
-  let indentLevel = 2;
-  if (typeof packageJson === 'string') {
-    wasString = true;
-    indentLevel = detectIndent(packageJson).indent;
-    if (packageJson.substr(-1) === '\n') {
-      endCharacters = '\n';
+const sortScripts = scripts => {
+  const prefixableScripts = defaultNpmScripts
```
Comment (fisker) [inline index.js]: @keithamus this should be a copy of `defaultNpmScripts` https://github.com/keithamus/sort-package-json/blob/bade9646d07d0c79059815701f2fdfde3e0e2f82/index.js#L109

**Comment by fisker** [comment]
Good to go.

**Pull request merged**